### PR TITLE
AZP: Run tests in Docker

### PR DESCRIPTION
**Pull request by gleon99**
Run CI in dockers

### CODE REVIEW

**Comment by yosefe** [comment]
@Artemy-Mellanox can you pls review?

File: buildlib/azure-pipelines-pr.yml [hunk 1]
```diff
@@ -231,6 +231,7 @@ stages:
   - stage: Distro
     dependsOn: [Static_check]
     jobs:
+# <<<<<<< HEAD
```
Comment (Artemy-Mellanox) [inline buildlib/azure-pipelines-pr.yml:234]: leftover?

File: src/uct/ib/base/ib_md.c [hunk 2]
```diff
@@ -1213,12 +1234,17 @@ static ucs_status_t uct_ib_query_md_resources(uct_component_t *component,
     }
 
     for (i = 0; i < num_devices; ++i) {
+        /* Skip non-existent and non-accessible devices */
+        if (!uct_ib_check_device_exists(device_list[i])) {
+            continue;
+        }
         ucs_snprintf_zero(resources[i].md_name, sizeof(resources[i].md_name),
```
Comment (Artemy-Mellanox) [inline src/uct/ib/base/ib_md.c:1241]: should it be something like `resources[num_resources].md_name`?

File: test/apps/test_ucp_dlopen.c [hunk 3]
```diff
@@ -56,12 +70,23 @@ int test_ucp_init(void *handle)
...
+        printf("Found component: %s\n", uct_components[i]->name);
+    }
+
+    uct_release_component_list_f(uct_components);
     ucp_cleanup_f(ucph);
 
     return 0;
 }
 
+
```
Comment (Artemy-Mellanox) [inline test/apps/test_ucp_dlopen.c:89]: leftover?

**Review by Artemy-Mellanox — DISMISSED** [review]

File: buildlib/docker_tests.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
+parameters:
```
Comment (yosefe) [inline buildlib/docker_tests.yml:1]: this file seems a duplication of tests.yml. what is the difference? can we reuse it and maybe just tweak some params?

File: contrib/test_jenkins.sh [hunk 5]
```diff
@@ -29,6 +29,11 @@ ucx_inst=${WORKSPACE}/install
 CUDA_MODULE="dev/cuda11.1.1"
 GDRCOPY_MODULE="dev/gdrcopy2.1_cuda11.1.1"
 
+if [[ "$RUNNING_IN_AZURE" == "yes" ]]; then
+	realdir=$(realpath $(dirname $0))
+	source ${realdir}/../buildlib/az-helpers.sh
```
Comment (yosefe) [inline contrib/test_jenkins.sh:34]: currently, whoever runs this script is responsible to source az-helpers before
see https://github.com/openucx/ucx/blob/4967de66bdab506c32e788574729cc3ec944c845/buildlib/tests.yml#L28

File: contrib/test_jenkins.sh [hunk 6]
```diff
@@ -72,6 +77,9 @@ MAKE="make"
 MAKEP="make -j${parallel_jobs}"
 export AUTOMAKE_JOBS=$parallel_jobs
 
+have_ib=$(ibv_devinfo -vv | grep GID || echo "")
```
Comment (yosefe) [inline contrib/test_jenkins.sh:80]: why need empty echo?

File: contrib/test_jenkins.sh [hunk 7]
```diff
@@ -72,6 +77,9 @@ MAKE="make"
...
+have_ptrace=$(capsh --print | grep Current | grep ptrace || echo "")
```
Comment (yosefe) [inline contrib/test_jenkins.sh:81]: 1. wrong grep, it is always false
```
$ capsh --print | grep Current
Current: =
$ capsh --print
Current: =
Bounding set =cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,35,36
Securebits: 00/0x0/1'b0
 secure-noroot: no (unlocked)
 secure-no-suid-fixup: no (unlocked)
 secure-keep-caps: no (unlocked)

```

2. use `capsh --supports=cap_sys_ptrace`

File: contrib/test_jenkins.sh [hunk 8]
```diff
@@ -880,6 +888,11 @@ test_profiling() {
 }
 
 test_ucs_load() {
+	if [ -z have_ptrace ]
```
Comment (yosefe) [inline contrib/test_jenkins.sh:891]: `${have_ptrace}`
this way it's always false

File: contrib/test_jenkins.sh [hunk 9]
```diff
@@ -880,6 +888,11 @@ test_profiling() {
...
+	then
+		echo "==== Not running UCS library loading test ===="
+		return
+	fi
```
Comment (yosefe) [inline contrib/test_jenkins.sh:895]: space line after

File: contrib/test_jenkins.sh [hunk 10]
```diff
@@ -880,6 +888,11 @@ test_profiling() {
 }
 
 test_ucs_load() {
+	if [ -z have_ptrace ]
+	then
+		echo "==== Not running UCS library loading test ===="
```
Comment (yosefe) [inline contrib/test_jenkins.sh:893]: add az_log_warning that test is not run because cma requires ptrace

File: contrib/test_jenkins.sh [hunk 11]
```diff
@@ -1278,9 +1290,12 @@ run_tests() {
 	$MAKEP
 	$MAKEP install
 
+if [ -n have_ib ]
+then
 	do_distributed_task 2 4 run_ucx_tl_check
-	do_distributed_task 1 4 run_ucp_hello
 	do_distributed_task 2 4 run_uct_hello
+fi
```
Comment (yosefe) [inline contrib/test_jenkins.sh:1297]: why these tests required ib? can we fix them to work w/o IB as well?

File: src/uct/ib/base/ib_md.c [hunk 12]
```diff
@@ -1185,6 +1185,27 @@ static uct_md_ops_t UCS_V_UNUSED uct_ib_md_global_odp_ops = {
     .detect_memory_type = ucs_empty_function_return_unsupported,
 };
 
+int uct_ib_check_device_exists(struct ibv_device *device)
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:1188]: int uct_ib_device_is_accessible(struct ibv_device *device)

File: src/uct/ib/base/ib_md.c [hunk 13]
```diff
@@ -1185,6 +1185,27 @@ static uct_md_ops_t UCS_V_UNUSED uct_ib_md_global_odp_ops = {
...
+{
+    /* Enough place to hold the full path */
+    char device_path[IBV_SYSFS_NAME_MAX + 16];
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:1191]: can we use `device->dev_path`?

File: src/uct/ib/base/ib_md.c [hunk 14]
```diff
@@ -1213,12 +1234,17 @@ static ucs_status_t uct_ib_query_md_resources(uct_component_t *component,
     }
 
     for (i = 0; i < num_devices; ++i) {
+        /* Skip non-existent and non-accessible devices */
+        if (!uct_ib_check_device_exists(device_list[i])) {
+            continue;
+        }
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:1240]: space line after

File: src/uct/ib/base/ib_md.c [hunk 15]
```diff
@@ -1193,7 +1214,7 @@ static ucs_status_t uct_ib_query_md_resources(uct_component_t *component,
     uct_md_resource_desc_t *resources;
     struct ibv_device **device_list;
     ucs_status_t status;
-    int i, num_devices;
+    int num_resources = 0, i, num_devices;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:1217]: num_resources should be first var, or init it before the for loop

File: src/uct/sm/mm/posix/mm_posix.c [hunk 16]
```diff
@@ -115,8 +116,17 @@ static size_t uct_posix_iface_addr_length(uct_mm_md_t *md)
 static ucs_status_t uct_posix_md_query(uct_md_h tl_md, uct_md_attr_t *md_attr)
 {
     uct_mm_md_t *md = ucs_derived_of(tl_md, uct_mm_md_t);
+    struct statvfs shm_statvfs;
+
+    if (statvfs(UCT_POSIX_SHM_OPEN_DIR, &shm_statvfs) < 0) {
```
Comment (yosefe) [inline src/uct/sm/mm/posix/mm_posix.c:121]: take dir path from configuration
https://github.com/openucx/ucx/blob/1fd6922226460fa6613f1e4fb2b8eeda67e1731a/src/uct/sm/mm/posix/mm_posix.c#L52

File: test/apps/test_ucp_dlopen.c [hunk 17]
```diff
@@ -117,4 +142,3 @@ int main(int argc, char **argv)
     printf("done\n");
     return ret;
 }
-
```
Comment (yosefe) [inline test/apps/test_ucp_dlopen.c:120]: pls restore

File: src/uct/ib/base/ib_md.c (hunk 2 above)
Comment (gleon99) [reply src/uct/ib/base/ib_md.c:1241]: A nice one! Thanks.

File: src/uct/ib/base/ib_md.c (hunk 13 above)
Comment (gleon99) [reply src/uct/ib/base/ib_md.c:1191]: I'm not sure about that.

`device->dev_path` is a symlink to the infiniband class device, which is a dir (e.g `/sys/class/infiniband_verbs/uverbs0 -> ../../devices/pci0000:00/0000:00:02.0/0000:04:00.0/infiniband_verbs/uverbs0`). It exists even in docker, and doesn't allow to determine, whether a device is accessible (unlike using `/dev/infiniband/uverbs0`).

Comment (yosefe) [reply src/uct/ib/base/ib_md.c:1191]: i see
maybe reuse IBV_SYSFS_PATH_MAX ?
or use dynamic string_buffer?

File: contrib/test_jenkins.sh (hunk 7 above)
Comment (gleon99) [reply contrib/test_jenkins.sh:81]: AFAIU, `--supports` is **not** what we need here. It tests whether the kernel version supports a particular capability in general, not whether it is indeed available at the moment.

Indeed, when running it inside docker, I get a positive result for ptrace, which is false.

Comment (yosefe) [reply contrib/test_jenkins.sh:81]: rightm supports is not. maybe grep for "Bounding set ="?

**Comment by gleon99** [comment]
@yosefe CR comments fixed, pipeline passed. Please review

File: contrib/test_jenkins.sh [hunk 18]
```diff
@@ -888,11 +882,12 @@ test_profiling() {
 }
 
 test_ucs_load() {
-	if [ -z have_ptrace ]
+	if [ -z "${have_ptrace}" ]
 	then
-		echo "==== Not running UCS library loading test ===="
+		log_warning "==== Not running UCS library loading test ===="
```
Comment (yosefe) [inline contrib/test_jenkins.sh:887]: azure_log_warning
https://github.com/openucx/ucx/blob/6769bd348b0543c25affbd9e6f760a37ecd3a9d1/buildlib/az-helpers.sh#L50

File: contrib/test_jenkins.sh [hunk 19]
```diff
@@ -937,11 +932,16 @@ test_ucp_dlopen() {
 
 test_init_mt() {
 	echo "==== Running multi-thread init ===="
+	# Each thread requires 5MB. Cap threads number by total available shared memory.
+	shm_available=$(df | grep shm | awk '{print $4}')
+	max_threads=$(($shm_available/5000))
```
Comment (yosefe) [inline contrib/test_jenkins.sh:937]: minor:
`max_threads=$(df /dev/shm | awk '/shm/ {printf "%d", $4 / 5000}')`

File: contrib/test_jenkins.sh [hunk 20]
```diff
@@ -937,11 +932,16 @@ test_ucp_dlopen() {
...
+	export OMP_NUM_THREADS=$(($max_threads < $(nproc) ? $max_threads : $(nproc)))
 	$MAKEP
 	for ((i=0;i<50;++i))
 	do
 		$AFFINITY timeout 1m ./test/apps/test_init_mt
 	done
+	unset OMP_NUM_THREADS
```
Comment (yosefe) [inline contrib/test_jenkins.sh:944]: minor: can set threads just for the one command
```
num_threads=$(($max_threads < $(nproc) ? $max_threads : $(nproc)))
for ...
do
    OMP_NUM_THREADS=$num_threads $AFFINITY timeout 1m ./test/apps/test_init_mt
done
```

File: src/uct/ib/base/ib_md.c [hunk 21]
```diff
@@ -1235,10 +1236,11 @@ static ucs_status_t uct_ib_query_md_resources(uct_component_t *component,
 
     for (i = 0; i < num_devices; ++i) {
         /* Skip non-existent and non-accessible devices */
-        if (!uct_ib_check_device_exists(device_list[i])) {
+        if (!uct_ib_device_is_accessible(device_list[i])) {
             continue;
         }
-        ucs_snprintf_zero(resources[i].md_name, sizeof(resources[i].md_name),
+
+        ucs_snprintf_zero(resources[num_resources].md_name, sizeof(resources[num_resources].md_name),
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:1243]: line seems too long

File: src/uct/sm/mm/posix/mm_posix.c [hunk 22]
```diff
@@ -116,9 +116,11 @@ static size_t uct_posix_iface_addr_length(uct_mm_md_t *md)
 static ucs_status_t uct_posix_md_query(uct_md_h tl_md, uct_md_attr_t *md_attr)
 {
     uct_mm_md_t *md = ucs_derived_of(tl_md, uct_mm_md_t);
+    const uct_posix_md_config_t *posix_config =
```
Comment (yosefe) [inline src/uct/sm/mm/posix/mm_posix.c:119]: align = with prev line

File: test/gtest/ucp/test_ucp_stream.cc [hunk 23]
```diff
@@ -564,6 +568,11 @@ UCS_TEST_P(test_ucp_stream, send_exp_recv_32) {
 
 UCS_TEST_P(test_ucp_stream, send_exp_recv_64) {
     ucp_datatype_t datatype = ucp_dt_make_contig(sizeof(uint64_t));
+    const uct_md_attr_t *md_attr = ucp_ep_md_attr(sender().ep(), 0);
+
+    if (has_transport("shm") && md_attr->cap.max_alloc < UCS_GBYTE) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_stream.cc:573]: add ( )

File: test/gtest/ucs/test_twheel.cc [hunk 24]
```diff
@@ -62,6 +64,17 @@ void twheel::timer_func(ucs_wtimer_t *self)
     t->self->timer_expired(t);
 }
 
+int twheel::check_all_timers_triggered(std::vector<struct hr_timer> &t)
+{
+    for (int i = 0; i < t.size(); i++) {
+        if(t[i].end_time == (ucs_time_t)0) {
```
Comment (yosefe) [inline test/gtest/ucs/test_twheel.cc:70]: - space after if
- return bool instead of int

File: test/gtest/ucs/test_twheel.cc [hunk 25]
```diff
@@ -229,13 +242,20 @@ UCS_TEST_F(twheel, delayed_sweep) {
...
 
-    sleep(1);
+    /* Sleep up to to seconds. Finish the test if the timers are triggered
+       earlier, fail otherwise. */
+    for (int i = 0; i < 10; i++) {
+        sleep(1);
+        ucs_twheel_sweep(&m_wheel, ucs_get_time());
 
-    ucs_twheel_sweep(&m_wheel, ucs_get_time());
+        if (!check_all_timers_triggered(t)) {
```
Comment (yosefe) [inline test/gtest/ucs/test_twheel.cc:251]: if (check_all_timers_triggered(t)) {
  return;
}

File: buildlib/azure-pipelines-pr.yml [hunk 26]
```diff
@@ -268,59 +268,6 @@ stages:
...
-              ./bin/ucx_perftest -t tag_bw &
-              server_pid=$!
-              sleep 1;
-              ./bin/ucx_perftest -t tag_bw localhost 2>&1 | tee perf.txt
-              wait $server_pid
-              grep "Final:" perf.txt
-            displayName: Test ucx_perftest
-            workingDirectory: $(System.DefaultWorkingDirectory)/ucx_bin_$(Build.BuildId)/install
-            env:
-              LD_LIBRARY_PATH: $(System.DefaultWorkingDirectory)/ucx_bin_$(Build.BuildId)/install/lib:$LD_LIBRARY_PATH
```
Comment (yosefe) [inline buildlib/azure-pipelines-pr.yml:323]: 👍

**Comment by gleon99** [comment]
@yosefe , fixed.

File: test/gtest/ucs/test_twheel.cc [hunk 27]
```diff
@@ -63,16 +63,15 @@ void twheel::timer_func(ucs_wtimer_t *self)
     struct hr_timer *t = ucs_container_of(self, struct hr_timer, timer);
     t->self->timer_expired(t);
 }
-
-int twheel::check_all_timers_triggered(std::vector<struct hr_timer> &t)
+bool twheel::check_all_timers_triggered(std::vector<struct hr_timer> &t)
```
Comment (yosefe) [inline test/gtest/ucs/test_twheel.cc:66]: minor: restore spaceline before

Comment (gleon99) [reply test/gtest/ucs/test_twheel.cc:67]: Fixed

**Comment by yosefe** [comment]
@gleon99 pls also fix conflict

**Comment by gleon99** [comment]
@yosefe  fixed.

**Comment by yosefe** [comment]
@gleon99 please use merge commit for resolving conflicts and do not rebase

**Comment by gleon99** [comment]
@yosefe Replaced with merge commit instead of the rebase.

**Review by yosefe — APPROVED** [review]
thanks!

File: contrib/test_jenkins.sh (hunk 18 above)
Comment (gleon99) [reply contrib/test_jenkins.sh:887]: I think it's better to revert to `log_warning` - it uses `azure_log_warning` if needed, and in other places in this file `log_warning` is used.

**Review by yosefe — APPROVED** [review]

**Comment by gleon99** [comment]
Squashed

**Pull request merged**